dali/operators/input/video_input.cc

```diff
@@ -40,23 +40,33 @@ VideoInput<Backend, FramesDecoder>::SetNextDataIdTrace(Workspace &ws, std::strin
 }
 
 
-template<>
-bool VideoInput<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
-                                       const Workspace &ws) {
+template<typename Backend, typename FramesDecoder>
+void VideoInput<Backend, FramesDecoder>::CreateDecoder(const Workspace &ws) {
+  auto sample = encoded_video_[0];
+  auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
+  size_t size = sample.shape().num_elements();
+  if constexpr (is_cpu) {
+    this->frames_decoders_[0] = std::make_unique<dali::FramesDecoder>(data, size, false);
+  } else {
+    this->frames_decoders_[0] = std::make_unique<FramesDecoderGpu>(data, size, ws.stream(), false);
+  }
+}
+
+
+template<typename Backend, typename FramesDecoder>
+bool VideoInput<Backend, FramesDecoder>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                                   const Workspace &ws) {
   if (!initialized_) {
     if (needs_data_load_) {
-      InputOperator<CPUBackend>::HandleDataAvailability();
-      LoadDataFromInputOperator(ws.GetThreadPool());
+      InputOperator<Backend>::HandleDataAvailability();
+      auto &tp = GetThreadPool(ws);
+      LoadDataFromInputOperator(tp);
     }
 
-    // Creating FramesDecoders
-    auto sample = encoded_video_[0];
-    auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
-    size_t size = sample.shape().num_elements();
-    frames_decoders_[0] = std::make_unique<FramesDecoder>(data, size, false);
+    CreateDecoder(ws);
 
     assert(output_descs_.empty());
-    DetermineOutputDescs(static_cast<int>(frames_decoders_[0]->NumFrames()));
+    DetermineOutputDescs(static_cast<int>(this->frames_decoders_[0]->NumFrames()));
```

Comment:
do you need all of those `this->`? I see you don't use it for other members?

Reply:
`this->` is needed, because the `frames_decoders_` is a member in the base class of the `VideoInput` (i.e. `VideoDecoderBase`). Maybe to be more elegant I can go with `using VideoDecoderBase::frames_decoders_`.